### PR TITLE
Cancel / Resubmit buttons in TxStatus

### PR DESCRIPTION
**Pull request by blurpesec**

### CODE REVIEW

File: src/components/GeneralStepper/QueryBanner.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  tooltipTranslationKey: string;
+}) => (
+  <BannerContainer>
+    <BannerContent>
+      <img src={runningIcon} />
+      <BannerText>{translate(contentTranslationKey)}</BannerText>
+      <Tooltip type={IconID.question} tooltip={translateRaw(tooltipTranslationKey)} />
+    </BannerContent>
+    <Link to={ROUTE_PATHS.DASHBOARD.path}>
+      <CancelBtn>CANCEL</CancelBtn>
```
Comment (FrederikBolding) [inline src/components/GeneralStepper/QueryBanner.tsx:59]: Translate

File: src/components/Query/WhenQueryExists.tsx [hunk 2]
```diff
@@ -1,44 +1,34 @@
...
   // This message banner is shown on the send form.
   // When passing in a complete transaction as query params the user will never reach the send form because the
   // stepper will slice out the first step UNLESS they don't have an associated network or account with the specified details
   // Therefore, we only display errors / messages.
   const deriveSendFormQueryWarning = (queries: IQueryResults) => {
     const queriesArePresent = Object.values(queries).some((v) => !!v);
-    const resubmitQueriesArePresent = MANDATORY_TRANSACTION_QUERY_PARAMS.every(
-      (resubmitParam) => queries[resubmitParam]
+    const txQueriesArePresent = MANDATORY_TRANSACTION_QUERY_PARAMS.every(
+      (speedupParam) => queries[speedupParam]
```
Comment (FrederikBolding) [inline src/components/Query/WhenQueryExists.tsx:20]: Can we rename to just param?

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 3]
```diff
@@ -182,6 +197,44 @@ export default function TxReceipt({
     }
   })();
 
+  const handleTxSpeedUpRedirect = async () => {
+    //dispatch({ type: txStatusReducer.actionTypes.TRIGGER_SPEED_UP });
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:201]: There is a bunch of commented out dispatches here?

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 4]
```diff
@@ -191,35 +244,53 @@ export default function TxReceipt({
     txConfig.network.id
   );
 
+  // cannot send from web3 or walletconnect wallets because they overwrite gas and nonce inputs.
+  const isSenderAccountPresent =
+    txConfig &&
+    accounts.find(
+      ({ address, wallet }) =>
+        isSameAddress(address, txConfig.senderAccount?.address) &&
+        ![
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:253]: Can we abstract this out to a helper? I think we already have one to detect web3

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 5]
```diff
@@ -231,12 +302,15 @@ export interface TxReceiptDataProps {
   sender: ISender;
   recipientContact: ExtendedContact | undefined;
   fiat: Fiat;
+  isSenderAccountPresent: StoreAccount | undefined;
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:305]: Rename since it's not a boolean.

File: src/features/SendAssets/SendAssets.reducer.tsx [hunk 6]
```diff
@@ -101,14 +102,14 @@ export const sendAssetsReducer = (state: State, action: ReducerAction): State =>
       return { ...state, send: false, txReceipt };
     }
 
-    case sendAssetsReducer.actionTypes.REQUEST_RESUBMIT: {
+    case sendAssetsReducer.actionTypes.REQUEST_SPEED_UP: {
```
Comment (FrederikBolding) [inline src/features/SendAssets/SendAssets.reducer.tsx:105]: Is this used at all?

File: src/features/TxStatus/helpers.ts [hunk 7]
```diff
@@ -33,11 +43,30 @@ export const fetchTxStatus = async ({
...
     config: fetchedTxConfig,
-    receipt: makeUnknownTxReceipt(txHash as ITxHash)(ITxType.UNKNOWN, fetchedTxConfig)
+    receipt:
+      fetchedTx && fetchedTx.confirmations
+        ? makeUnknownTxReceipt(txHash as ITxHash)(ITxType.UNKNOWN, fetchedTxConfig)
+        : makePendingTxReceipt(txHash as ITxHash)(ITxType.UNKNOWN, fetchedTxConfig)
+  };
+};
+
+export const createQueryParams = (txConfig: ITxConfig, type: TxQueryTypes) => {
```
Comment (FrederikBolding) [inline src/features/TxStatus/helpers.ts:59]: Test

File: src/features/SendAssets/helpers.spec.ts [hunk 8]
```diff
@@ -35,78 +35,80 @@ const validERC20ResubmitQuery = {
...
 
   it('fails to derive txConfig when invalid eth tx query parameters are included', () => {
-    const parsedQueryParams = parseQueryParams(invalidResubmitQuery)(
-      [fNetwork],
-      fAssets,
-      fAccounts
-    );
-    expect(parsedQueryParams).toStrictEqual({ type: 'resubmit', txConfig: undefined });
+    const parsedQueryParams = parseQueryParams(invalidSpeedUpQuery)([fNetwork], fAssets, fAccounts);
+    expect(parsedQueryParams).toStrictEqual({ type: TxQueryTypes.SPEEDUP, txConfig: undefined });
```
Comment (FrederikBolding) [inline src/features/SendAssets/helpers.spec.ts:72]: Add tests for cancel

File: src/features/SendAssets/SendAssets.reducer.tsx (hunk 6 above)
Comment (FrederikBolding) [reply src/features/SendAssets/SendAssets.reducer.tsx:105]: If so, are we using both the reducer and the query params?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

**Comment by FrederikBolding** [comment]
Needs overall more tests too

File: src/features/SendAssets/SendAssets.reducer.tsx (hunk 6 above)
Comment (blurpesec) [reply src/features/SendAssets/SendAssets.reducer.tsx:105]: not used anymore, i used a find&replace to rename :P ill delete tho

**Comment by FrederikBolding** [comment]
Needs a rebase too!

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 9]
```diff
@@ -182,6 +193,20 @@ export default function TxReceipt({
     }
   })();
 
+  const handleTxSpeedUpRedirect = async () => {
+    if (!txConfig) return;
+    const { fast } = await fetchGasPriceEstimates(txConfig.network);
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:198]: Error handling?

File: src/components/TransactionFlow/TxReceipt.tsx [hunk 10]
```diff
@@ -182,6 +193,20 @@ export default function TxReceipt({
...
+    const query = constructSpeedUpTxQuery(txConfig, fast);
+    history.replace(`${ROUTE_PATHS.SEND.path}/?${query}`);
+  };
+
+  const handleTxCancelRedirect = async () => {
+    if (!txConfig) return;
+    const { fast } = await fetchGasPriceEstimates(txConfig.network);
```
Comment (FrederikBolding) [inline src/components/TransactionFlow/TxReceipt.tsx:205]: Error handling?

File: src/database/versions.ts [hunk 11]
```diff
@@ -2,7 +2,7 @@ import moment from 'moment';
 
 import { SETTINGS_FILENAME } from '@config';
 import { LocalStorage } from '@types';
-import { noOp } from '@utils';
+import { noOp } from '@utils/noOp';
```
Comment (FrederikBolding) [inline src/database/versions.ts:5]: Import from `@utils`

File: src/features/SendAssets/helpers.spec.ts [hunk 12]
```diff
@@ -35,78 +35,147 @@ const validERC20ResubmitQuery = {
...
+  it('fails to derive txConfig when there is no added account with from address - speed up', () => {
+    const parsedQueryParams = parseQueryParams(validETHSpeedUpQuery)([fNetwork], fAssets, []);
+    expect(parsedQueryParams).toStrictEqual({ type: TxQueryTypes.SPEEDUP, txConfig: undefined });
   });
 });
 
-describe('parseResubmitParams', () => {
-  it('correctly parses valid erc20 resubmit query params', () => {
-    const parsedResubmitParams = parseResubmitParams(validERC20ResubmitQuery)(
+describe('parseTransactionQueryParams', () => {
```
Comment (FrederikBolding) [inline src/features/SendAssets/helpers.spec.ts:153]: No tests for cancel here?

File: src/features/SendAssets/helpers.ts [hunk 13]
```diff
@@ -107,22 +108,27 @@ export const parseQueryParams = (queryParams: any) => (
...
       return {
-        type: 'resubmit',
-        txConfig: parseResubmitParams(queryParams)(networks, assets, accounts)
+        type: TxQueryTypes.SPEEDUP,
+        txConfig: parseTransactionQueryParams(queryParams)(networks, assets, accounts)
+      };
+    case TxQueryTypes.CANCEL:
+      return {
+        type: TxQueryTypes.CANCEL,
+        txConfig: parseTransactionQueryParams(queryParams)(networks, assets, accounts)
```
Comment (FrederikBolding) [inline src/features/SendAssets/helpers.ts:119]: Can we reuse this across the two cases?

File: src/types/asset.tsx [hunk 14]
```diff
@@ -25,17 +31,13 @@ export interface Asset {
   readonly contractAddress?: string;
   readonly decimal?: number;
   readonly isCustom?: boolean;
+  readonly mappings?: IMappings;
```
Comment (FrederikBolding) [inline src/types/asset.tsx:34]: Are you sure this is available on all assets and not just `ExtendedAsset`?

File: src/utils/wallets.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
+import { WalletId, StoreAccount, TAddress } from '@types';
+import { isSameAddress, isWeb3Wallet } from '@utils';
+
+export const isViewOnlyWallet = (walletId: WalletId): boolean => walletId === WalletId.VIEW_ONLY;
+
+export const isSenderAccountPresent = (
```
Comment (FrederikBolding) [inline src/utils/wallets.ts:6]: Can we rename it to make it clear that it isn't just detecting an account but a specific type? Maybe with a comment too? Can we reflect it in the prop name?

File: src/utils/queries.test.ts [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+      'type=speedup&from=0xB2BB2b958aFA2e96dAb3F3Ce7162B87dAea39017&to=0xB2BB2b958aFA2e96dAb3F3Ce7162B87dAea39017&gasLimit=0x5208&nonce=0x6&chainId=3&value=0x2386f26fc10000&data=0x&gasPrice=0x2e90edd000';
+    const testFastGasPrice = 200;
+    const txQuery = constructSpeedUpTxQuery(fETHNonWeb3TxConfig, testFastGasPrice);
+    expect(txQuery).toEqual(expectedSpeedUpTxQuery);
+  });
+});
+
+describe('createQueryParamsDefaultObject', () => {
+  it('correctly constructs a speed up tx param object for an erc20 transfer', () => {
+    const zed = createQueryParamsDefaultObject(fERC20NonWeb3TxConfig, TxQueryTypes.SPEEDUP);
```
Comment (FrederikBolding) [inline src/utils/queries.test.ts:47]: zed?

**Review by FrederikBolding — CHANGES_REQUESTED** [review]

File: src/components/TransactionFlow/TxReceipt.tsx (hunk 9 above)
Comment (blurpesec) [reply src/components/TransactionFlow/TxReceipt.tsx:198]: it's handled by that function already :)

File: src/components/TransactionFlow/TxReceipt.tsx (hunk 10 above)
Comment (blurpesec) [reply src/components/TransactionFlow/TxReceipt.tsx:205]: same as above.

File: src/database/versions.ts (hunk 11 above)
Comment (blurpesec) [reply src/database/versions.ts:5]: Ill check this again, but it was causing a circular dependency issue with the `@utils/queries.ts` file

File: src/features/SendAssets/helpers.spec.ts (hunk 12 above)
Comment (blurpesec) [reply src/features/SendAssets/helpers.spec.ts:153]: `cancel` and `speedup` both use the same transaction query format. They just extend it with `type: 'cancel' | 'speedup'` to tell the interface which banners/errors to display

File: src/features/SendAssets/helpers.ts (hunk 13 above)
Comment (blurpesec) [reply src/features/SendAssets/helpers.ts:119]: yup, read above ^

File: src/utils/wallets.ts (hunk 15 above)
Comment (blurpesec) [reply src/utils/wallets.ts:11]: I just called it main type for now :-) We can look into using toggles on the different wallet types, but i think that will involve a fair bit of refactoring and doesn't need to be done now
image

**Review by FrederikBolding — APPROVED** [review]
lgtm

**Pull request merged**